extensions/amp-analytics/0.1/events.js

```diff
@@ -436,34 +704,105 @@ export class TimerEventTracker extends EventTracker {
     const timerSpec = config['timerSpec'];
     user().assert(timerSpec && typeof timerSpec == 'object',
         'Bad timer specification');
-    user().assert('interval' in timerSpec,
-        'Timer interval specification required');
-    const interval = Number(timerSpec['interval']) || 0;
-    user().assert(interval >= MIN_TIMER_INTERVAL_SECONDS,
-        'Bad timer interval specification');
-    const maxTimerLength = 'maxTimerLength' in timerSpec ?
-        Number(timerSpec['maxTimerLength']) : DEFAULT_MAX_TIMER_LENGTH_SECONDS;
-    user().assert(maxTimerLength == null || maxTimerLength > 0,
-        'Bad maxTimerLength specification');
-    const callImmediate = 'immediate' in timerSpec ?
-        Boolean(timerSpec['immediate']) : true;
-
-    const win = this.root.ampdoc.win;
-    const intervalId = win.setInterval(() => {
-      listener(this.createEvent_(eventType));
-    }, interval * 1000);
-    this.trackers_.push(intervalId);
-    win.setTimeout(() => {
-      this.removeTracker_(intervalId);
-    }, maxTimerLength * 1000);
-    if (callImmediate) {
-      listener(this.createEvent_(eventType));
+    const timerStart = 'startSpec' in timerSpec ? timerSpec['startSpec'] : null;
+    user().assert(!timerStart || typeof timerStart == 'object',
+        'Bad timer start specification');
+    const timerStop = 'stopSpec' in timerSpec ? timerSpec['stopSpec'] : null;
+    user().assert((!timerStart && !timerStop) || typeof timerStop == 'object',
+        'Bad timer stop specification');
+
+    const timerId = this.generateTimerId_();
+    let startBuilder;
+    let stopBuilder;
+    if (timerStart) {
+      const startTracker = this.getTracker_(timerStart);
+      user().assert(startTracker, 'Cannot track timer start');
+      startBuilder = startTracker.add.bind(startTracker, context,
+          timerStart['on'], timerStart,
+          this.handleTimerToggle_.bind(this, timerId, eventType, listener));
     }
+    if (timerStop) {
+      const stopTracker = this.getTracker_(timerStop);
+      user().assert(stopTracker, 'Cannot tracker timer stop');
+      stopBuilder = stopTracker.add.bind(stopTracker, context,
```

Comment:
One question: What should we do with invalid `timerStart`/`timerStop` spec. Especially with invalid `timerStop` spec? Never stop or stop the timer with error. WDYT? I am OK we address this in separate PR, but will need to finalize the behavior before adding documentation.

Reply:
Well, when the tracker is added it will run through the usual validations inside the tracker#add method. timerStart will therefore be immediately validated upon page load, but timerStop would not be validated until a start event happens, causing a user-frustrating delay. (As is, this would mean a timer with an invalid stop config would therefore never stop.)

Really, I think the ideal would be to move that validation forward. This would require refactoring all of the trackers to either provide a #validate method to be used in absence of a call to #add, or possibly simply taking validation out of the trackers entirely and doing it from the amp root upon instantiation (also allows natural validation for non-trigger configs like requests and transport in the same place).

Either way, definitely a big change, so I want to know that this is a very impactful change before we decide to do it. If we have the convention that users are expected to test their code, would a descriptive error after timer start be sufficient? If so, there may not be a point to fixing such a large systemic issue.

Reply:
I totally get your point!

Sure, the perfect solution is to move validation forward to the very beginning. We don't have that before because all trigger events are independent of each other, and validate at runtime is good enough. 

I don't expect we make this big refactor soon. My question was: with invalid `timerStop`,  (with `user.assert()` or `user.error()`) should we stop the timer right away, or should we never stop. I personally prefer to stop timer right away. 

I am imaging code to be something like
```
try {
  stopBuilder=stopTracker.add(....)
} catch {
  //Stop Timer. 
}
```

Reply:
I've switched it so it will dispose all listeners and log an error with an invalid stop spec.